### ISSUE TITLE
Interpolation during resize does not use premultiplied alpha

### ISSUE THREAD

**Issue by antmelnyk** [image-rs/image#1655]
The issue happens when using the `image::imageops::sample::resize` function with any filter type, for example, `FilterType::Gaussian`.
When resizing the image that has abrupt edges and transparency around it, a [classic issue](https://ciechanow.ski/alpha-compositing/#filtering) happens and RGB values from transparency are leaking, creating a border around the image. On the screenshot, you can see a few rotated squares resized and their edges having black borders because there are `Rgba([0, 0, 0, 0])` pixels that got leaked during interpolation.

<img width="1184" alt="Screenshot 2022-01-20 at 03 13 47" src="https://user-images.githubusercontent.com/17108334/150244521-c7601d60-1453-41ef-9897-8d64a3d67a74.png">

I assume initially crate authors knew about the issue, because there is premultiplied alpha used in `Blend` trait implementation for `Rgba` pixel, but seems like there isn't something similar implemented for sampling during resizing. I tried to add pre multiplying during sampling and it fixed the problem.

In `imageops/sample.rs` before when applying weight to the pixel I did this:
```rust
       let a = vec.3 / 255.0;
                     
      // Multiply each channel on alpha
      (
        t.0 + (vec.0 * a) * w,
        t.1 + (vec.1 * a) * w,
        t.2 + (vec.2 * a) * w,
        t.3 + vec.3 * w,
      )
    });

    // Divide each channel by alpha if it's not 0
    let t: P = Pixel::from_channels(
      NumCast::from(FloatNearest(clamp(if t.0 == 0.0 { t.0 } else { t.0 / (t.3 / 255.0)  }, min, max))).unwrap(),
      NumCast::from(FloatNearest(clamp(if t.1 == 0.0 { t.1 } else { t.1 / (t.3 / 255.0)   }, min, max))).unwrap(),
      NumCast::from(FloatNearest(clamp(if t.2 == 0.0 { t.2 } else { t.2 / (t.3 / 255.0)   }, min, max))).unwrap(),
      NumCast::from(FloatNearest(clamp(t.3, min, max))).unwrap(),
    );
```

<img width="886" alt="Screenshot 2022-01-20 at 03 14 49" src="https://user-images.githubusercontent.com/17108334/150244525-37db52f8-024d-4de1-bd13-80f64918b3f2.png">

The question is, am I understanding/doing something wrong or there was some reason for the existing implementation to be as it is? Clearly, the fix is working for me, but I'd like to hear any suggestions from crate maintainers who for sure know more about image processing than me 😃

**Comment by fintelia** [comment]
A lot of the code in `imageops` hasn't been changed in years, though we've been gradually fixing bugs in it as they've been discovered. If your code is producing the desired results then I'd trust it more than whatever we have implemented

**Comment by antmelnyk** [comment]
Fair enough 🥲 Initial function was written 7 years ago and wasn't changed conceptually since then as far as I see.

Without fixing the sampling function itself, it works if I perform pre-multiplying before passing buffer inside the `resize` and then de-multiplying it again after. Not as efficient (2 more iterations over image pixels), but it is what it is. It probably would make sense to add an additional parameter to `resize`, something like `premultiplied: bool` as it is handled [in some other](https://sharp.pixelplumbing.com/api-composite) image processing APIs, or just add new function.

**Comment by awused** [comment]
You can pretty easily pull the resample function out and use it on ImageBuffers while still using the rest of the image crate. I've personally modified it to resample in linear light with premultiplied alpha in parallel (a pretty naive port of the code to rayon gives close enough to linear speedup), though I don't think it'd be suitable as the default implementation.

**Comment by drewcassidy** [comment]
If this is added, it should be optional. there are cases where premultiplication is not desired, such as in 3d graphics applications where alpha represents something other than transparency
